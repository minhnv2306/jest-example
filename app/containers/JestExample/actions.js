/*
 *
 * JestExample actions
 *
 */

import { DEFAULT_ACTION, LOAD_USER_SAGA, LOADED_USER } from './constants';

export function defaultAction() {
	return {
		type: DEFAULT_ACTION,
	};
}

export function loadUser() {
	return {
		type: LOAD_USER_SAGA,
	};
}

export function loadedUser(user) {
	return {
		type: LOADED_USER,
		user,
	};
}
