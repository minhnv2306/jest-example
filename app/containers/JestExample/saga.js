import { call, put, takeLatest } from 'redux-saga/effects';
import { LOAD_USER_SAGA } from 'containers/JestExample/constants';
import { loadedUser } from 'containers/JestExample/actions';

import request from 'utils/request';

/**
 * Github repos request/response handler
 */
export function* loadUser() {
  const requestURL = `https://jsonplaceholder.typicode.com/todos/1`;

  try {
    // Call our request helper (see 'utils/request')
    const user = yield call(request, requestURL);
    console.log('Hello Saga');
    yield put(loadedUser(user));
  } catch (err) {
    console.log(err);
    // yield put(repoLoadingError(err));
    console.log(1);
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* loadUserFromFakeAPI() {
  // Watches for LOAD_REPOS actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_USER_SAGA, loadUser);
}
