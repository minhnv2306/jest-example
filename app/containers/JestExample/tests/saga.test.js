/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { put } from 'redux-saga/effects';
import { loadUser } from '../saga';
import { loadedUser } from '../actions';

// const generator = jestExampleSaga();

const user = {
  name: 'MinhNV',
};
describe('jestExampleSaga Saga', () => {
  let loadUserGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    loadUserGenerator = loadUser();

    const selectDescriptor = loadUserGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    const response = {
      userId: 1,
      id: 1,
      title: 'delectus aut autem',
      completed: false,
    };
    const putDescriptor = loadUserGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(loadedUser(response, user)));
  });
});
