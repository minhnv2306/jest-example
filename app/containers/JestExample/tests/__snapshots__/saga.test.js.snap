// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jestExampleSaga Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://jsonplaceholder.typicode.com/todos/1",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`jestExampleSaga Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/JestExample/LOADED_USER",
      "user": Object {
        "name": "MinhNV",
      },
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
